zerver/migrations/0423_fix_email_gateway_attachment_owner.py

```diff
@@ -0,0 +1,176 @@
+                    has_attachment=False,
+                    realm_id=attachment.realm_id,
+                    sending_client_id=internal_client.id,
+                    date_sent__gte=attachment.create_time,
+                    date_sent__lte=attachment.create_time + timedelta(minutes=5),
+                    content__contains="/user_uploads/" + attachment.path_id,
+                ).order_by("date_sent")
+            )
+        if len(possible_matches) == 0:
+            print("  No matches!")
+            continue
+
+        # If there are 1 or more matches, we assume the earliest is
+        # the correct one, since it's ~impossible to have predicted
+        # the URL before it was first sent.
+        message = possible_matches[0]
+        print(f"  Found {message.id} @ {message.date_sent} by {message.sender.delivery_email})")
+
+        # If this is an ArchivedMessage, then we have to move the
+        # Attachment into an ArchivedAttachment.  We also have to
+        # generate an zerver_archivedattachment_message row with an id
+        # based on the next free from zerver_attachment_message, since
+        # those are one id space.
+        if isinstance(message, ArchivedMessage):
+            # move_rows
+            fields = list(Attachment._meta.fields)
+            src_fields = [Identifier("zerver_attachment", field.column) for field in fields]
+            dst_fields = [Identifier(field.column) for field in fields]
+            with connection.cursor() as cursor:
+                raw_query = SQL(
+                    """
+                    INSERT INTO zerver_archivedattachment ({dst_fields})
+                        SELECT {src_fields}
+                        FROM zerver_attachment
+                        WHERE id = {id}
+                    ON CONFLICT (id) DO NOTHING
+                    RETURNING id
+                    """
+                )
+                cursor.execute(
+                    raw_query.format(
+                        src_fields=SQL(",").join(src_fields),
+                        dst_fields=SQL(",").join(dst_fields),
+                        id=Literal(attachment.id),
+                    )
+                )
+                archived_ids = [id for (id,) in cursor.fetchall()]
+                if len(archived_ids) != 1:
+                    print("!!! Did not create one archived attachment row!")
+            attachment.delete()
+            attachment = ArchivedAttachment.objects.get(id=archived_ids[0])
+
+        # Determine message (and thus attachment) properties; this is
+        # from do_claim_attachments
+        is_message_realm_public = False
+        is_message_web_public = False
+        if message.recipient.type == 2:  # Recipient.STREAM
+            stream = Stream.objects.get(id=message.recipient.type_id)
+            is_message_realm_public = not stream.invite_only and not stream.is_in_zephyr_realm
+            is_message_web_public = stream.is_web_public
```

Comment:
I guess this ends up using the current values of these settings, rather than what they were at the time the message was sent. This is fine, because we reset these to `None` in order to recalculate them when moving messages or editing stream permissions; see for example, `validate_attachment_request_for_spectator_access`. But we could just set these both to `None` safely and not have to duplicate the calculation here if we preferred.

(As a side note, we should probably add a bonus commit to fix the `AbstractAttachment` comments to explain what it means to set them to `None`).

Reply:
It doesn't seem hard to set these correctly, and I don't think this logic is going to drift over time.

I'll file a separate PR for the None value.

Reply:
This already looks doc'd:
```python3
    # The two fields below serve as caches to let us avoid looking up
    # the corresponding messages/streams to check permissions before
    # serving these files.
    #
    # For both fields, the `null` state is used when a change in
    # message permissions mean that we need to determine their proper
    # value.

    # Whether this attachment has been posted to a public stream, and
    # thus should be available to all non-guest users in the
    # organization (even if they weren't a recipient of a message
    # linking to it).
    is_realm_public = models.BooleanField(default=False, null=True)
    # Whether this attachment has been posted to a web-public stream,
    # and thus should be available to everyone on the internet, even
    # if the person isn't logged in.
    is_web_public = models.BooleanField(default=False, null=True)

Reply:
OK, I just didn't read closely enough. Sounds good!